README.md
# portfolio_MonomatchGame
online game adaptation of the »Dobble« card game. \
https://www.dobblegame.com/

Inspired by: \
»How does Dobble (Spot It) work?« (Stand-up Maths) \
https://www.youtube.com/watch?v=VTDKqW_GLkw  \
and \
»How to make the card game Dobble (and the Maths behind it!)« (MirKat Maths) \
https://www.youtube.com/watch?v=oyqD1Sg5M4Q  \

The Fano plane algorithm by Micky Dore: \
https://mickydore.medium.com/the-dobble-algorithm-b9c9018afc52

I choose to have cards showing 8 numbers each - which makes a total number of 57 different cards. \

The game: \
Start the game and you'll see 2 cards on the screen. \
Try to find the only element that is present on both of the cards. \
The background of the number turns orange if you have clicked on the correct element. \
Click »New card set« for a new set of cards. 

Hint: With every reload of the page you'll get either a set of numbers or icons that are either upright or rotate randomly. 

Enjoy :)

Continuous integration via netlify.com: \
https://portfolio-martin442-monomatchgame.netlify.app

CHANGELOG.md
# 1.3.1
12-07-2021
- minor adjustments

# 1.3
12-07-2021
- Start screen added
- icons added: array of 57 fontawesome.com icons 
- random rotation to numbers/icons added
- random value to select either rotation/no rotation or icons/numbers on each page reload
- rotated numbers (6, 9, 16, 18, 19, 51) are underlined to avoid confusion
- Element ID's removed from CSS to destinct between CSS and JS - according to this hint: https://csswizardry.com/2013/05/hashed-classes-in-css/ 
- a couple of class-name changes to cover possible future features…


# 1.2.1
01-07-2021
- QR-code added and formatted (URL: portfolio-martin442-monomatchgame.netlify.app)
- »New game« button placed over the cardset for better mobile landscape orientation experience
- »New game« z-index/position added to .fadein and .fadeout

# 1.2
28-06-2021
- »New game« button added
- »New game« button appears if dobble is clicked on one card
- opponents cards transforms to opacity 

# 1.1 
23-06-2021
- CHANGELOG.md added
- color set adjusted
- google font added
- <a> changed to <button>
- add <span> inside <button> to better adjust font position for different fonts 

# 1.0
14-06-2021
Initial feature-set: 
- MirKat Maths "theDoubleAlgorithm.js" creates an array of 57 elements, containing an array of 8 numbers each
- two circles on the screen representing two cards with 8 numbers on each card
- an eventListener "onClick" on each number allows the user to click on any number on both cards
- a click on a number changes the font color
- each click on a number checks if that number is included in the set of numbers of the other card
- while clicking on a number that is present on both cards, the background of the number changes to green and all eventListeners are removed

scripts/faIcons.js
const fe_icons_57 = [
    '<i class="fab fa-accessible-icon"></i>', 
    '<i class="far fa-address-book"></i>',
    '<i class="fas fa-adjust"></i>', 
    '<i class="fas fa-allergies"></i>',
    '<i class="fas fa-ambulance"></i>',
    '<i class="fas fa-anchor"></i>',
    '<i class="fab fa-angellist"></i>',
    '<i class="fas fa-archive"></i>',
    '<i class="fas fa-arrow-circle-down"></i>',
    '<i class="fas fa-at"></i>',
    '<i class="fas fa-band-aid"></i>', 
    '<i class="fas fa-bath"></i>',
    '<i class="fas fa-basketball-ball"></i>',
    '<i class="fas fa-bed"></i>',
    '<i class="far fa-bell"></i>',
    '<i class="fas fa-bell-slash"></i>',
    '<i class="fas fa-bicycle"></i>',
    '<i class="fas fa-biohazard"></i>',
    '<i class="fas fa-birthday-cake"></i>',
    '<i class="fas fa-bomb"></i>',
    '<i class="fas fa-briefcase-medical"></i>',
    '<i class="fas fa-bullhorn"></i>',
    '<i class="fas fa-campground"></i>',
    '<i class="fas fa-camera-retro"></i>',
    '<i class="fas fa-candy-cane"></i>',
    '<i class="fas fa-car"></i>',
    '<i class="fas fa-carrot"></i>',
    '<i class="fas fa-cat"></i>',
    '<i class="fas fa-chair"></i>',
    '<i class="fas fa-chart-line"></i>',
    '<i class="fas fa-cheese"></i>',
    '<i class="fas fa-child"></i>',
    '<i class="fas fa-clipboard-list"></i>',
    '<i class="fas fa-cloud"></i>',
    '<i class="fas fa-coffee"></i>',
    '<i class="far fa-comment-dots"></i>',
    '<i class="far fa-compass"></i>',
    '<i class="fas fa-couch"></i>',
    '<i class="fas fa-crow"></i>',
    '<i class="fas fa-cubes"></i>',
    '<i class="fas fa-desktop"></i>',
    '<i class="fas fa-dice-five"></i>',
    '<i class="fas fa-dog"></i>',
    '<i class="fas fa-dove"></i>',
    '<i class="far fa-edit"></i>',
    '<i class="far fa-envelope"></i>',
    '<i class="fas fa-faucet"></i>',
    '<i class="far fa-file-image"></i>',
    '<i class="fas fa-fire-extinguisher"></i>',
    '<i class="fas fa-fish"></i>',
    '<i class="far fa-flag"></i>',
    '<i class="fab fa-fort-awesome-alt"></i>',
    '<i class="fas fa-gamepad"></i>',
    '<i class="fas fa-glass-cheers"></i>',
    '<i class="far fa-grin-alt"></i>',
    '<i class="fas fa-guitar"></i>',
    '<i class="far fa-hand-peace"></i>',
    '<i class="far fa-heart"></i>'    
]

scripts/theDoubleAlgorithm.js
// @author: Mickey Dore 
// https://mickydore.medium.com/the-dobble-algorithm-b9c9018afc52
// MirKat Maths youtube channel: 
// "How to make the card game Double (and the Maths behind it!)" 
// https://www.youtube.com/watch?v=oyqD1Sg5M4Q

let i, j, k
let n = 7                 //order of plane, must be a prime number
let numOfSymbols = n + 1  //order of plane + 1
let cards = [] //the deck of cards
let card = []; //the current card we are building
//to start, we build the first card
for (i = 1; i<= n+1; i++) {
    card.push(i)
}
cards.push(card)
//then we build the next n number of cards
for (j=1; j<=n; j++) {
    card = []
    card.push(1)
    
    for (k=1; k<=n; k++) {
        card.push(n * j + (k+1))
    }
    cards.push(card)
}
//finally we build the next n² number of cards
for (i= 1; i<=n; i++) {
    for (j=1; j<=n; j++) {
        card = []
        card.push(i+1)
        
        for (k=1; k<= n; k++) {
            card.push(n+2+n*(k-1)+(((i-1)*(k-1)+j-1)%n))
        }
        cards.push(card)
    }
}


scripts/main.js
// @author Martin Schwedes
// game logic on screen

const cardsOnScreen = [
    '#card1',
    '#card2'
];
let cardSet1 = [];
let cardSet2 = [];

// fwrxh array "cards" from theDoubleAlgorithm.js
const gameCards = cards;

// create template for number buttons
const buttonsOnCardTemplate = document.querySelector('._gameset_zen button');
// define parts of gameplay
const welcomeCard = document.querySelector('._gamestart');
const GameZenCard = document.querySelector('._gameset_zen');
// define button elements
const newGameButton = welcomeCard.querySelector('._startgame');
const newCardsButton = document.querySelector('._newcards');

// store opponents cards
let opponent = '';
let angle = '(0deg)';
const underLineChar = [6, 9, 16, 18, 19, 51]; 
let allowRotation = false; 
let allowIcons = true; 

/* random set of rotation/no rotation and icons/numbers */
allowRotation = Math.floor(Math.random()*2); 
allowIcons = Math.floor(Math.random()*2); 


initGameScreen(); 
// initZenMode();

function initGameScreen() {
    welcomeCard.style.display = 'flex';
    newGameButton.classList.add('font__Kalem'); 
    welcomeCard.querySelector('h1').classList.add('font__Kalem'); 
    welcomeCard.querySelector('p').classList.add('font__Kalem'); 
    GameZenCard.style.display = 'none';
    newCardsButton.style.display = 'none';
    newGameButton.addEventListener('click', startNewZenGame );
}

function shuffleNums(a) {
    let j, x, i;
    for (i = a.length - 1; i > 0; i--) {
        j = Math.floor(Math.random() * (i + 1));
        x = a[i];
        a[i] = a[j];
        a[j] = x;
    }
    return a;
}

function initZenMode() {
    // let new game button disappear
    newCardsButton.style.display = 'block';
    newCardsButton.classList.add('font__Kalem'); 
    newCardsButton.classList.add('fadeout');
    // remove all button elements from screen
    const removeButtons = document.querySelectorAll('._gameset_zen button');
    for (const i of removeButtons) {
        i.remove();
    }

    // loop through cards in gameset (2 cards)
    for (const oneCardOnScreen of cardsOnScreen) {
        // random card from card stack
        const randomCard = Math.floor(Math.random() * gameCards.length);
        // shuffle numbers in card array
        const randomCardNums = shuffleNums(gameCards[randomCard]);
        // delete random card from array to avoid same cards in the gameset
        gameCards.splice(randomCard,1);
        // create 2 cardsets defined outside the loop
        if (oneCardOnScreen == cardsOnScreen[0]) {
            cardSet1 = randomCardNums;
        } else {
            cardSet2 = randomCardNums;
        }

        // loop through numbers for one card
        // loop through numbers on card to create a visual card set
        for (const number of randomCardNums) {
            const buttonOnCard = buttonsOnCardTemplate.cloneNode(true);
            const buttonOnCardText = buttonOnCard.querySelector('span');
            document.querySelector(oneCardOnScreen).append(buttonOnCard);
            buttonOnCardText.classList.add('cardNr'+number);
            if (allowIcons) { /* icons or unbers */
                buttonOnCardText.innerHTML = fe_icons_57[number];
            } else {
                buttonOnCardText.innerText = number;
            }
            buttonOnCardText.classList.add('font__Kalem');
            if (allowRotation) { /* rotate for more fun :) */ 
                angle = '(' + (Math.floor(Math.random() * 5)) * 72 + "deg)";
                buttonOnCardText.style.transform = 'rotate' + angle;
                if (underLineChar.includes(number)) {
                    buttonOnCardText.style.textDecoration = 'underline';
                    buttonOnCardText.style.paddingTop = '0.1rem';
                }
            }
        }

        // add eventListener on each number
        const numbersOnCards = document.querySelectorAll(oneCardOnScreen + ' button span');
        for (const eachNumA of numbersOnCards) {
            eachNumA.addEventListener('click', checkNum);
        }
    }
}

function checkNum(event) {
    const clickedClasses = this.classList; 
    const clickedNum = clickedClasses[0].substr(6, clickedClasses[0].length - 6); /* number from className, ie 'cardNr23' */
    const clickedNumParent = this.parentElement.parentElement;
    const clickedNumElement = document.querySelector('#' + clickedNumParent.id + ' .cardNr' + clickedNum);
    clickedNumElement.style.color = 'var(--click-number)';
    let matchingNumElement = '';
    if (clickedNumParent.id == 'card1') {
        matchingNumElement = document.querySelector('#card2 .cardNr' + clickedNum);
        if (cardSet2.includes(parseInt(clickedNum))) {
            clickedNumElement.parentElement.style.backgroundColor = 'var(--click-success)';
            opponent = matchingNumElement.parentElement;
            opponent.parentElement.classList.add('success');
            opponent.style.backgroundColor = 'var(--click-success)';
            endTheGame();
        }
    } else if (clickedNumParent.id == 'card2') {
        matchingNumElement = document.querySelector('#card1 .cardNr' + clickedNum);
        if (cardSet1.includes(parseInt(clickedNum))) {
            clickedNumElement.parentElement.style.backgroundColor = 'var(--click-success)';
            opponent = matchingNumElement.parentElement;
            opponent.parentElement.classList.add('success');
            opponent.style.backgroundColor = 'var(--click-success)';
            endTheGame(); 
        }
    }
}

function startNewZenGame() {
    welcomeCard.style.display = 'none';
    GameZenCard.style.display = 'flex';
    initZenMode(); 
}

function startNextZenGame() {
    newCardsButton.removeEventListener('click', initZenMode );
    newCardsButton.classList.toggle('fadein');
    opponent.parentElement.classList.remove('success');
    initZenMode(); 
}

function endTheGame() {
    // removing all eventListeners for the numbers
    const removeEL = document.querySelectorAll('button span')
    for (const i of removeEL) {
        i.removeEventListener('click', checkNum);
    }
    /* show new game button */
    newCardsButton.classList.add('fadein');
    newCardsButton.addEventListener('click', startNextZenGame );
}



// to do: 
// - timer
// - 2 player mode